File: src/main/java/com/datascience/project/PlaylistMapper.java
/**
 * Author: Abuchi Okeke
 * Date: 04/28/2020
 * Mapper class for mapReduce algorithm to find matching
 * songs using items from user inputs
 */

package com.datascience.project;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Mapper;

import java.io.IOException;
import java.util.regex.Pattern;

public class PlaylistMapper extends Mapper<LongWritable, Text, Text, NullWritable> {
     
   /*Correct the name of the class "FindPlaylistsJobKt" for the kotlin byte code
    *Or call direct using @JvmStatic or @JvmField (add to the companion object and call class direct)
    *Not sure if it's going to work */

    String userArtistName = UserInput.artist;
    String userTrackName = UserInput.track;
    String userKeyWord = UserInput.keyword;
  
    private static final Pattern WORD_BOUNDARY = Pattern.compile(",");  
 
    //Method to find matching songs
    boolean checkUserInputExistence(String userinput, String spotifyinput) {
    String input = userinput.toLowerCase();
    String spotifyItem = spotifyinput.toLowerCase();

    return spotifyItem.matches(".*" + input + ".*");
}
    public void map(LongWritable offset, Text lineText, Context context)
        throws IOException, InterruptedException {

       //split data by comma ","
       String[] line = WORD_BOUNDARY.split(lineText.toString());
       //Get spotify items
   String artistname = line[1];
   String trackname = line[2];
   String playlistname = line[3];
  
      if (userKeyWord != null){
      //boolean isFound = playlistnames.indexOf(userKeyWord) !=-1? true: false;
      }
          // Matthew Russell: bug fix to write more relevant playlist names
          if (checkUserInputExistence(userArtistName, artistname) && checkUserInputExistence(userKeyWord, playlistname)
                  && (userTrackName.isEmpty() || checkUserInputExistence(userTrackName, trackname))){
              context.write(new Text(playlistname), NullWritable.get());
        }

    }
  }

File: src/main/kotlin/com/datascience/project/Playlist_Mapper.kt
package com.datascience.project

import org.apache.hadoop.conf.Configured
import org.apache.hadoop.fs.Path
import org.apache.hadoop.io.IntWritable
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.NullWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.mapreduce.Job
import org.apache.hadoop.mapreduce.Mapper
import org.apache.hadoop.mapreduce.Reducer
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat
import org.apache.hadoop.mapreduce.lib.output.MultipleOutputs
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat
import org.apache.hadoop.util.Tool

class Playlist_Mapper: Mapper<LongWritable, Text, NullWritable, Text>() {
    override fun map(offset: LongWritable, line: Text, context: Context) {
        val lineStr = line.toString()
        val sortedLine = lineStr
            .split(',')
        // Load the file as string of Tract, Artist, Playlist.
        // Split the string using a delimiter (,).
        // Output the playlist name..

        context.write(NullWritable.get(), Text(sortedLine[2]))
    }
}